data/provider/src/main/java/tech/pegasys/teku/api/stateselector/StateSelectorFactory.java

```diff
@@ -79,20 +85,45 @@ public StateSelector genesisSelector() {
 
   @Override
   public StateSelector finalizedSelector() {
+
     return () ->
         SafeFuture.completedFuture(
             client
                 .getLatestFinalized()
                 .map(
-                    finalized ->
-                        addMetaData(
-                            finalized.getState(),
-                            // The finalized checkpoint may change because of optimistically
-                            // imported blocks at the head and if the head isn't optimistic, the
-                            // finalized block can't be optimistic.
-                            client.isChainHeadOptimistic(),
-                            true,
-                            true)));
+                    finalized -> {
+                      final UInt64 slot = finalized.getSlot();
+                      final Spec spec = finalized.getSpec();
+                      final BeaconState finalizedState;
+                      //if this isn't the first slot after the finalized epoch than
+                      // we'll compute empty slots until we get there
+                      if (!slot.dividedBy(spec.getSlotsPerEpoch(slot)).isZero()) {
```

Comment:
I'd avoid checks like this especially we consider changes in slots time or maybe slots per epoch in the future. Use spec helpers like `computeStartSlotAtEpoch` and compare actual and expected slot

Reply:
yeah, that is much better, thanks for the heads up.